File: apps/mcp-server/src/tools/shopping/adapters/base-adapter.ts
import {
  Product,
  ProductSearchOptions,
  ProductSearchResult,
  CartItem,
  Cart,
  ShoppingOperationResult,
  SiteAdapterNameValues
} from "@shopping-copilot/shared";
import { WebsiteConfig } from "../types";

/**
 * Abstract base class for shopping website adapters
 * Each website implementation must extend this class
 */
export abstract class BaseShoppingAdapter {
  protected website: SiteAdapterNameValues;
  protected config: WebsiteConfig;

  constructor(
    website: SiteAdapterNameValues,
    config: WebsiteConfig
  ) {
    this.website = website;
    this.config = config;
  }

  /**
   * Search for products on the website
   */
  abstract searchProducts(
    options: ProductSearchOptions,
    credentials: any
  ): Promise<ShoppingOperationResult<ProductSearchResult>>;

  /**
   * Add a product to the shopping cart
   */
  abstract addToCart(
    productId: string,
    quantity: number,
    credentials: any,
    variant?: string
  ): Promise<ShoppingOperationResult<CartItem | string>>;

  /**
   * Remove an item from the shopping cart
   */
  abstract removeFromCart(
    cartItemId: string,
    credentials: any
  ): Promise<ShoppingOperationResult<boolean>>;

  /**
   * Update quantity of an item in the cart
   */
  abstract updateCartQuantity(
    cartItemId: string,
    quantity: number,
    credentials: any
  ): Promise<ShoppingOperationResult<CartItem>>;

  /**
   * Get current cart contents
   */
  abstract getCartContents(credentials: any): Promise<ShoppingOperationResult<Cart>>;

  /**
   * Validate that the adapter is properly configured
   */
  protected validateConfig(): boolean {
    if (!this.config) {
      return false;
    }

    // Note: Authentication validation is done per-method call
    // since credentials are passed as parameters, not stored in the adapter
    return true;
  }

  /**
   * Get website name for identification
   */
  getWebsiteName(): SiteAdapterNameValues {
    return this.website;
  }

  /**
   * Get rate limit information
   */
  getRateLimit(): number {
    return this.config.rateLimitPerMinute;
  }

  /**
   * Create standardized error result
   */
  protected createErrorResult<T>(error: string): ShoppingOperationResult<T> {
    return {
      success: false,
      error,
      website: this.website,
    };
  }

  /**
   * Create standardized success result
   */
  protected createSuccessResult<T>(data: T): ShoppingOperationResult<T> {
    return {
      success: true,
      data,
      website: this.website,
    };
  }

  /**
   * Sanitize product data to remove potential security issues
   */
  protected sanitizeProduct(product: any): Product {
    return {
      id: String(product.id || '').replace(/[<>]/g, ''),
      title: String(product.title || '').replace(/[<>]/g, '').slice(0, 200),
      description: String(product.description || '').replace(/[<>]/g, '').slice(0, 1000),
      price: Number(product.price) || 0,
      currency: String(product.currency || 'USD').slice(0, 3),
      imageUrl: this.sanitizeUrl(product.imageUrl),
      availability: Boolean(product.availability),
      rating: product.rating ? Math.min(Math.max(Number(product.rating), 0), 5) : undefined,
      reviewCount: product.reviewCount ? Math.max(Number(product.reviewCount), 0) : undefined,
      category: product.category ? String(product.category).replace(/[<>]/g, '').slice(0, 100) : undefined,
      brand: product.brand ? String(product.brand).replace(/[<>]/g, '').slice(0, 100) : undefined,
      url: this.sanitizeUrl(product.url),
    };
  }

  /**
   * Sanitize URLs to prevent injection attacks
   */
  private sanitizeUrl(url?: string): string | undefined {
    if (!url) return undefined;
    
    try {
      const parsed = new URL(url);
      // Only allow http/https protocols
      if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
        return undefined;
      }
      return parsed.toString();
    } catch {
      return undefined;
    }
  }
}

File: apps/python-backend/src/tools/shopping/adapters/base_adapter.py
"""
Abstract base class for shopping website adapters
Port of the TypeScript BaseShoppingAdapter class
"""

from abc import ABC, abstractmethod
from typing import Optional, Union
import re
from urllib.parse import urlparse

from ..constants import SiteAdapterName
from ..types import (
    Product, ProductSearchOptions, ShoppingOperationResult, WebsiteConfig, CredentialsType
)
from ..security import ShoppingSecurity


class BaseShoppingAdapter(ABC):
    """Abstract base class for shopping website adapters"""
    
    def __init__(self, website: SiteAdapterName, config: WebsiteConfig):
        self.website = website
        self.config = config
    
    @abstractmethod
    async def search_products(
        self,
        options: ProductSearchOptions,
        credentials: CredentialsType
    ) -> ShoppingOperationResult:
        """Search for products on the website"""
        pass
    
    @abstractmethod
    async def add_to_cart(
        self,
        product_id: str,
        quantity: int,
        credentials: CredentialsType,
        variant: Optional[str] = None
    ) -> ShoppingOperationResult:
        """Add a product to the shopping cart"""
        pass
    
    @abstractmethod
    async def remove_from_cart(
        self,
        cart_item_id: str,
        credentials: CredentialsType
    ) -> ShoppingOperationResult:
        """Remove an item from the shopping cart"""
        pass
    
    @abstractmethod
    async def update_cart_quantity(
        self,
        cart_item_id: str,
        quantity: int,
        credentials: CredentialsType
    ) -> ShoppingOperationResult:
        """Update quantity of an item in the cart"""
        pass
    
    @abstractmethod
    async def get_cart_contents(self, credentials: CredentialsType) -> ShoppingOperationResult:
        """Get current cart contents"""
        pass
    
    def validate_config(self) -> bool:
        """Validate that the adapter is properly configured"""
        if not self.config:
            return False
        
        # Note: Authentication validation is done per-method call
        # since credentials are passed as parameters, not stored in the adapter
        return True
    
    def get_website_name(self) -> SiteAdapterName:
        """Get website name for identification"""
        return self.website
    
    def get_rate_limit(self) -> int:
        """Get rate limit information"""
        return self.config.rate_limit_per_minute
    
    def create_error_result(self, error: str) -> ShoppingOperationResult:
        """Create standardized error result"""
        return ShoppingOperationResult(
            success=False,
            error=ShoppingSecurity.format_secure_error(error),
            website=self.website.value
        )
    
    def create_success_result(self, data) -> ShoppingOperationResult:
        """Create standardized success result"""
        return ShoppingOperationResult(
            success=True,
            data=data,
            website=self.website.value
        )
    
    def sanitize_product(self, product_data: dict) -> Product:
        """Sanitize product data to remove potential security issues"""
        return Product(
            id=str(product_data.get('id', '')).replace('<', '').replace('>', '')[:100],
            title=str(product_data.get('title', '')).replace('<', '').replace('>', '')[:200],
            description=str(product_data.get('description', '')).replace('<', '').replace('>', '')[:1000],
            price=float(product_data.get('price', 0)) if product_data.get('price') is not None else 0.0,
            currency=str(product_data.get('currency', 'ILS'))[:3],
            image_url=self._sanitize_url(product_data.get('image_url')),
            availability=bool(product_data.get('availability', True)),
            rating=self._sanitize_rating(product_data.get('rating')),
            review_count=max(int(product_data.get('review_count', 0)), 0) if product_data.get('review_count') is not None else None,
            category=str(product_data.get('category', '')).replace('<', '').replace('>', '')[:100] if product_data.get('category') else None,
            brand=str(product_data.get('brand', '')).replace('<', '').replace('>', '')[:100] if product_data.get('brand') else None,
            # barcode=str(product_data.get('barcode', '')).replace('<', '').replace('>', '')[:100] if product_data.get('barcode') else None,
            url=self._sanitize_url(product_data.get('url'))
        )
    
    def _sanitize_url(self, url: Optional[str]) -> Optional[str]:
        """Sanitize URLs to prevent injection attacks"""
        if not url:
            return None
        
        try:
            parsed = urlparse(url)
            # Only allow http/https protocols
            if parsed.scheme not in ('http', 'https'):
                return None
            return parsed.geturl()
        except Exception:
            return None
    
    def _sanitize_rating(self, rating) -> Optional[float]:
        """Sanitize rating value"""
        if rating is None:
            return None
        try:
            rating_float = float(rating)
            return max(0.0, min(5.0, rating_float))
        except (ValueError, TypeError):
            return None